toolkit/components/filewatcher/NativeFileWatcherMac.h

```diff
@@ -0,0 +1,255 @@
+    CallBackEvents()
+        : callBackLock("filewatcher::eventLock")
+    {}
+    std::queue<CallBackAction> mSavedEvents; // Used to hold individual FSEvents
+    Mutex callBackLock; // Used for safe access across multiple threads.
+};
+
+class NativeFileWatcherIOTask; // Forward Declaration of class.
+
+
+/**
+ * The NativeFileWatcherFSETask class which holds all of the functionality of the FSEventStream.
+ *
+ * It inherits from Runnable as we need to make a blocking call to CFRunLoopRun(). CFRunLoopRun() allows the
+ * stream to report file system events. The event stream is an immutable object, therefore with every change in the
+ * paths watched, the stream must be stopped and a new stream must be created with a new list of paths to be watched.
+ */
+class NativeFileWatcherFSETask : public Runnable
+{
+public:
+    explicit NativeFileWatcherFSETask(NativeFileWatcherIOTask* parent, CallBackEvents* cbe, std::vector<CFStringRef>& dirs);
+
+    // This is a blocking method which runs the steam on a CFRunLoop.
+    NS_IMETHOD Run() override;
+
+    // Adds path to the stream (and restarts it)
+    NS_IMETHOD AddPath(char* pathToAdd);
+
+    // Removes path from the stream and restarts it. Unless it removes the last path being watched by the stream.
+    // In this instance, we will return from this runnable without starting a new stream.
+    NS_IMETHOD RemovePath(char* pathToRemove);
+
+private:
+    //mozilla::Mutex runLoopLock;
+
+    FSEventStreamRef mEventStreamRef; // Reference to the FSEventStream
+    CFRunLoopRef mRunLoop = nullptr; // Reference to the RunLoop which is passed back to the parent class, to stop the RunLoop
+    std::vector<CFStringRef> mDirs; // A mutable list of directories which we build dynamically based on the needs of the stream
+    static CallBackEvents* cbe_internal; // Pointer to a callback events struct which is allocated in the parent class and used
+                                         // to hold events read from the stream for processing back in the parent class.
+    NativeFileWatcherIOTask* mParent; // Pointer to the parent class to allow stopping the CFRunLoop from another thread.
+                                      // Please see the variable mRunLoop.
+
+    // Callback which handles events from the stream and is automatically called from the running stream
+    // which we have set up to monitor the user defined paths.
+    static void fsevents_callback(ConstFSEventStreamRef streamRef,
+                                               void *clientCallBackInfo,
+                                               size_t numEvents,
+                                               void *eventPaths,
+                                               const FSEventStreamEventFlags eventFlags[],
+                                               const FSEventStreamEventId eventIds[]);
+
+    // Utility function which gets all currently watched paths from the stream except for skip, if supplied.

```

Comment:
Looks like you're going to allocate a new vector each time you call this method. Is this really needed?

Reply:
me to review.

Reply:
We need a mutable list of paths when we restart the stream and want to remove a path being watched, so it'll either be a vector or some other list implementation. It is stack allocation, and that could be a problem with very long watch lists, although I don't know what kind of stack limit modern platforms have. Would you prefer it to be heap allocated, or have some other method we didn't think of?

Reply:
> It is stack allocation

Vectors are pretty much always heap-allocated. Am I missing something?

Right now, I don't have a clear enough idea of how this function works and when it's called to offer an alternative, but it seems a bit odd to heap-allocate each time.